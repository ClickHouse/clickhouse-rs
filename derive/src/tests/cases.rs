//! # Snapshot tests for `Row` derive macro.
//!
//! - This module contains happy path tests. For failures, see `tests/ui/`.
//! - These tests are supposed to be used as intermediate checks for complex
//!   code generation, not as a final integration tests.

use super::render;

#[test]
fn simple_owned_row() {
    render! {
        #[derive(Row)]
        struct Sample {
            a: i32,
            b: String,
        }
    }
}

#[test]
fn generic_owned_row() {
    render! {
        #[derive(Row)]
        struct Sample<T> {
            a: i32,
            b: T,
        }
    }

    render! {
        #[derive(Row)]
        struct Sample<A, B> {
            a: A,
            b: B,
        }
    }

    render! {
        #[derive(Row)]
        struct Sample<T> where T: Clone {
            a: i32,
            b: T,
        }
    }
}

#[test]
fn simple_borrowed_row() {
    render! {
        #[derive(Row)]
        struct Sample<'a> {
            a: i32,
            b: &'a str,
        }
    }
}

#[test]
fn generic_borrowed_row() {
    render! {
        #[derive(Row)]
        struct Sample<'a, T> {
            a: i32,
            b: &'a T,
        }
    }

    render! {
        #[derive(Row)]
        struct Sample<'a, A, B> {
            a: A,
            b: &'a B,
        }
    }

    render! {
        #[derive(Row)]
        struct Sample<'a, T> where T: Clone {
            a: i32,
            b: &'a T,
        }
    }
}

#[test]
fn serde_rename() {
    render! {
        #[derive(Row)]
        struct Sample {
            a: i32,
            #[serde(rename = "items.a")]
            items_a: Vec<String>,
            #[serde(rename = "items.b")]
            items_b: Vec<u32>,
        }
    }
}

#[test]
fn serde_skip_serializing() {
    render! {
        #[derive(Row)]
        struct Sample {
            a: u32,
            #[serde(skip_serializing)]
            b: u32,
        }
    }
}

#[test]
fn serde_skip_deserializing() {
    render! {
        #[derive(Row)]
        struct Sample {
            a: u32,
            #[serde(skip_deserializing)]
            b: u32,
        }
    }
}

#[test]
fn crate_attribute() {
    render! {
        #[derive(Row)]
        #[clickhouse(crate = "foo")]
        struct Foo {
            a: u32,
            b: u32,
        }
    }
}
