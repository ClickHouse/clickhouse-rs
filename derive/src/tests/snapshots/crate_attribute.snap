---
source: derive/src/tests/cases.rs
assertion_line: 125
---
#[derive(Row)]
#[clickhouse(crate = "foo")]
struct Foo {
    a: u32,
    b: u32,
}

/****** GENERATED ******/
#[automatically_derived]
impl foo::Row for Foo {
    const NAME: &'static str = stringify!(Foo);
    const COLUMN_NAMES: &'static [&'static str] = &["a", "b"];
    const COLUMN_COUNT: usize = <Self as foo::Row>::COLUMN_NAMES.len();
    const KIND: foo::_priv::RowKind = foo::_priv::RowKind::Struct;
    type Value<'__v> = Self;
}
